ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -0,0 +1,9955 @@
+|		IF_NOT_Z_TYPE, reg, IS_REFERENCE, >1
+|		GET_Z_PTR reg, reg
+|		add reg, offsetof(zend_reference, val)
+|1:
+||	}
+|.endmacro
+
+|.macro SAVE_VALID_OPLINE, op
+||	if (op == last_valid_opline) {
+|		mov aword EX->opline, IP
+||	} else {
+|		ADDR_OP2_2 mov, aword EX->opline, op, r0
+||	}
+|.endmacro
+
+// zval should be in FCARG1a
+|.macro ZVAL_DTOR_FUNC, var_info, opline // arg1 must be in FCARG1a
+||	do {
+||		if (has_concrete_type((var_info) & (MAY_BE_STRING|MAY_BE_ARRAY|MAY_BE_OBJECT|MAY_BE_RESOURCE))) {
+||			zend_uchar type = concrete_type((var_info) & (MAY_BE_STRING|MAY_BE_ARRAY|MAY_BE_OBJECT|MAY_BE_RESOURCE));
+||			if (type == IS_STRING && !ZEND_DEBUG) {
+|				test byte [FCARG1a + 5], IS_STR_PERSISTENT
+|				jnz >1
+|				EXT_CALL _efree, r0
+|				jmp >2
+|1:
+|				EXT_CALL free, r0
+|2:
+||				break;
+||			} else if (type == IS_ARRAY) {
+||				if (opline) {
+|					SAVE_VALID_OPLINE opline
+||				}
+|				EXT_CALL zend_array_destroy, r0
+||				break;
+||			} else if (type == IS_OBJECT) {
+||				if (opline) {
+|					SAVE_VALID_OPLINE opline
+||				}
+|				EXT_CALL zend_objects_store_del, r0
+||				break;
+||			}
+||		}
+||		if (opline) {
+|			SAVE_VALID_OPLINE opline
+||		}
+|		EXT_CALL rc_dtor_func, r0
+||	} while(0);
+|.endmacro
+
+|.macro ZVAL_PTR_DTOR, addr, op_info, gc, cold, safe, opline
+||	if ((op_info) & (MAY_BE_STRING|MAY_BE_ARRAY|MAY_BE_OBJECT|MAY_BE_RESOURCE|MAY_BE_REF)) {
+||		if ((op_info) & ((MAY_BE_ANY|MAY_BE_UNDEF)-(MAY_BE_OBJECT|MAY_BE_RESOURCE))) {
+|			// if (Z_REFCOUNTED_P(cv)) {
+||			if (cold) {
+|				IF_ZVAL_REFCOUNTED addr, >1
+|.cold_code
+|1:
+||			} else {
+|				IF_NOT_ZVAL_REFCOUNTED addr, >4
+||			}
+||		}
+|		// if (!Z_DELREF_P(cv)) {
+|		GET_ZVAL_PTR FCARG1a, addr
+|		GC_DELREF FCARG1a
+||		if (RC_MAY_BE_1(op_info)) {
+||			if (RC_MAY_BE_N(op_info)) {
+||				if (gc && RC_MAY_BE_N(op_info) && ((op_info) & (MAY_BE_REF|MAY_BE_ARRAY|MAY_BE_OBJECT))) {
+|					jnz >3
+||				} else {
+|					jnz >4
+||				}
+||			}
+||			if (safe) {
+|				// ZVAL_NULL(cv);
+|				SET_ZVAL_TYPE_INFO addr, IS_NULL
+||			}
+|			// zval_dtor_func(r);
+|			ZVAL_DTOR_FUNC op_info, opline
+||			if (gc && RC_MAY_BE_N(op_info) && ((op_info) & (MAY_BE_REF|MAY_BE_ARRAY|MAY_BE_OBJECT))) {
+|				jmp >4
+||			}
+|3:
+||		}
+||		if (gc && RC_MAY_BE_N(op_info) && ((op_info) & (MAY_BE_REF|MAY_BE_ARRAY|MAY_BE_OBJECT))) {
+||			if ((op_info) & MAY_BE_REF) {
+||				zend_jit_addr ref_addr = ZEND_ADDR_MEM_ZVAL(ZREG_FCARG1a, offsetof(zend_reference, val));
+|				IF_NOT_ZVAL_TYPE addr, IS_REFERENCE, >1
+|				IF_NOT_ZVAL_REFCOUNTED ref_addr, >4
+|				GET_ZVAL_PTR FCARG1a, ref_addr
+|1:
+||			}
+|			IF_GC_MAY_NOT_LEAK FCARG1a, eax, >4
+|			// gc_possible_root(Z_COUNTED_P(z))
+|			EXT_CALL gc_possible_root, r0
+||			if (cold && ((op_info) & ((MAY_BE_ANY|MAY_BE_UNDEF)-(MAY_BE_OBJECT|MAY_BE_RESOURCE)))) {
+|				jmp >4
+|.code
+||			}
+||		} else if (cold && ((op_info) & ((MAY_BE_ANY|MAY_BE_UNDEF)-(MAY_BE_OBJECT|MAY_BE_RESOURCE)))) {
+|			jmp >4
+|.code
+||		}
```

Comment:
This code is the same at the end of the if and in the else if, it should be fine to move it outside as a single check.

Reply:
By the way, I found the implementation of this macro quite confusing, it would probably be a lot more obvious if MAY_BE_N was a completely separate codepath, at the expense of duplicating some of the code.

Reply:
Let's delay optimizations until merge